src/lib/utils/concepts.h

```diff
@@ -29,6 +33,93 @@ struct is_strong_type<Strong<Ts...>> : std::true_type {};
 template <typename... Ts>
 constexpr bool is_strong_type_v = is_strong_type<std::remove_const_t<Ts>...>::value;
 
+namespace ranges {
+
+/**
+ * Models a std::ranges::contiguous_range that (optionally) restricts its
+ * value_type to ValueT. In other words: a stretch of contiguous memory of
+ * a certain type (optional ValueT).
+ */
+template <typename T, typename ValueT = std::ranges::range_value_t<T>>
+concept contiguous_range = std::ranges::contiguous_range<T> && std::same_as<ValueT, std::ranges::range_value_t<T>>;
+
+/**
+ * Models a std::ranges::contiguous_range that satisfies
+ * std::ranges::output_range with an arbitrary value_type. In other words: a
+ * stretch of contiguous memory of a certain type (optional ValueT) that can be
+ * written to.
+ */
+template <typename T, typename ValueT = std::ranges::range_value_t<T>>
+concept contiguous_output_range = contiguous_range<T, ValueT> && std::ranges::output_range<T, ValueT>;
+
+/**
+ * Models a range that can be turned into a std::span<>. Typically, this is some
+ * form of ranges::contiguous_range.
+ */
+template <typename T>
+concept spanable_range = std::constructible_from<std::span<const std::ranges::range_value_t<T>>, T>;
+
+/**
+ * Find the length in bytes of a given contiguous range @p r.
+ */
+inline constexpr size_t size_bytes(spanable_range auto&& r) {
+   return std::span{r}.size_bytes();
+}
+
+/**
+ * Check whether a given range type has a length known at compile time.
+ */
+template <spanable_range R>
+bool consteval has_static_extent() {
+   return decltype(std::span{std::declval<R>()})::extent != std::dynamic_extent;
+}
+
+/**
+ * Check that a given range @p r has a certain statically-known byte length. If
+ * the range's extent is known at compile time, this is a static check,
+ * otherwise a runtime argument check will be added.
+ *
+ * @throws Invalid_Argument  if range @p r has a dynamic extent and does not
+ *                           feature the expected byte length.
+ */
+template <size_t expected, spanable_range R>
+inline constexpr void assert_exact_byte_length(R&& r) {
+   const std::span s{r};
+   if constexpr(has_static_extent<R>()) {
+      static_assert(s.size_bytes() == expected, "memory region does not have expected byte lengths");
+   } else {
+      BOTAN_ARG_CHECK(s.size_bytes() == expected, "memory region does not have expected byte lengths");
+   }
+}
```

Comment:
It's interesting, because a call of this method works, but only with type deduction. For example:
```cpp
    std::array<uint8_t,32> a;
    // Does not work 
    assert_exact_byte_length<32, std::array<uint8_t,32>>(a);
    // Works
    assert_exact_byte_length<32>(a);
```
It seems the input type is implicitly converted to a span.

Reply:
That's somewhat expected, because of the forwarding reference (`R&&`). When passing `a` with type deduction, you'll end up with `R = std::array<>&` (an lvalue reference). I.e. If you were to add the `&` to your explicit declaration, it should work. Here's some reading: https://medium.com/factset/modern-c-in-depth-perfect-forwarding-570f242261f8